### PR TITLE
[js/web] WebAssembly profiling

### PR DESCRIPTION
**Pull request by hanbitmyths**
This PR enables profiling in WebAssembly. It simply prints out all profiling data into console. Using MEMFS or IndexedDB will take into consideration next time, which needs C API support and better understanding of WebAssembly file system.

### CODE REVIEW

File: js/web/lib/wasm/binding/ort-wasm.d.ts [hunk 1]
```diff
@@ -43,6 +43,8 @@ export interface OrtWasmModule extends EmscriptenModule {
   _OrtCreateRunOptions(logSeverityLevel: number, logVerbosityLevel: number, terminate: boolean, tag: number): number;
   _OrtAddRunConfigEntry(runOptionsHandle: number, configKey: number, configValue: number): number;
   _OrtReleaseRunOptions(runOptionsHandle: number): void;
+
+  _OrtEndProfiling(sessionHandle: number): number;
```
Comment (fs-eire) [inline js/web/lib/wasm/binding/ort-wasm.d.ts:47]: Do we need `_OrtStartProfiling` as well?

File: onnxruntime/core/common/profiler.h [hunk 2]
```diff
@@ -121,7 +121,16 @@ class Profiler {
   // Mutex controlling access to profiler data
   OrtMutex mutex_;
   bool enabled_{false};
+#if defined(__wasm__)
+  /*
+   * The simplest way to emit profiling data in WebAssembly is to print out to console,
+   * since browsers can't access to a file system directly.
+   * TODO: Consider MEMFS or IndexedDB instead of console.
+   */
+  std::ostream& profile_stream_{std::cout};
```
Comment (fs-eire) [inline onnxruntime/core/common/profiler.h:130]: is it a good idea to lazy-initialize? ( to save memory usage )

File: onnxruntime/wasm/api.h [hunk 3]
```diff
@@ -41,7 +41,7 @@ int EMSCRIPTEN_KEEPALIVE OrtInit(int num_threads, int logging_level);
  * @param enable_cpu_mem_arena enable or disable cpu memory arena
  * @param enable_mem_pattern enable or disable memory pattern
  * @param execution_mode sequential or parallel execution mode
- * @param enable_profiling enable or disable profiling. it's a no-op and for a future use.
+ * @param enable_profiling enable or disable profiling.
```
Comment (fs-eire) [inline onnxruntime/wasm/api.h:44]: is profiling per-session only?

File: js/web/lib/wasm/binding/ort-wasm.d.ts (hunk 1 above)
Comment (hanbitmyths) [reply js/web/lib/wasm/binding/ort-wasm.d.ts:47]: ORT C API doesn't support StartProfiling yet. Profiling is started when a session is created.

File: onnxruntime/core/common/profiler.h (hunk 2 above)
Comment (hanbitmyths) [reply onnxruntime/core/common/profiler.h:130]: I don't think we can do with std::cout unless we replace all profile_stream_ to std::cout explicitly, which would be ugly.

Comment (fs-eire) [reply onnxruntime/core/common/profiler.h:130]: is it possible to initialize profile_stream_ to std::cout at https://github.com/microsoft/onnxruntime/pull/8932/files#diff-d0251c9ed4b9395c8bacb3dc7d56d6421fd1e4c7ef8fd16683fc604272ad12dbR52 ?

File: onnxruntime/wasm/api.h (hunk 3 above)
Comment (hanbitmyths) [reply onnxruntime/wasm/api.h:44]: Native ORT can support multiple profiling in a session, but there is no exposed StartProfiling C API. Before that, it's per-session only now.

**Review by fs-eire — APPROVED** [review]

**Pull request merged**